modeling/local-slice-algorithm.go

```diff
@@ -0,0 +1,171 @@
+			endpointsAvailable.push(endpointDeviation{name: zoneName, deviation: deviation})
+			localGroup.Composition[zoneName] = int(expectedEndpoints)
+		} else {
+			endpointsNeeded.push(endpointDeviation{name: zoneName, deviation: -deviation})
+			localGroup.Composition[zoneName] = zone.Endpoints
+		}
+
+		sliceGroups[zoneName] = localGroup
+	}
+
+	err := balanceSliceGroups(&endpointsAvailable, &endpointsNeeded, sliceGroups)
+	if err != nil {
+		return nil, err
+	}
+	return sliceGroups, nil
+}
+
+// balanceSliceGroups helps to 'rent' endpoints from zones have extra endpoints
+// to EndpointSliceGroups that have insufficient ones
+func balanceSliceGroups(endpointsAvailable *endpointsList, endpointsNeeded *endpointsList, sliceGroups map[string]EndpointSliceGroup) error {
+	for _, receiveZone := range endpointsNeeded.byZone {
+		// The available list is empty while there are still endpoints in
+		// need. This can happen when the approximation on deviation
+		// (calculated above) ends up in asymmetric sums of endpoints
+		// available and endpoints in need (in need > available)
+		if len(endpointsAvailable.byZone) == 0 {
+			makeupForApproximation(endpointsNeeded, sliceGroups, false)
+			return nil
+		}
+		// This is ugly, but the index below is a copy, after pop no way to
+		// udpate the index to its new position
+		indexDelta := 0
+		for index, sendZone := range endpointsAvailable.byZone {
+			if sendZone.deviation == receiveZone.deviation {
+				sliceGroups[receiveZone.name].Composition[sendZone.name] = sendZone.deviation
+				endpointsAvailable.pop()
+				break
+			}
+			if sendZone.deviation > receiveZone.deviation {
+				sliceGroups[receiveZone.name].Composition[sendZone.name] = receiveZone.deviation
+				endpointsAvailable.byZone[index-indexDelta].deviation -= receiveZone.deviation
+				break
+			}
+			if sendZone.deviation < receiveZone.deviation {
+				sliceGroups[receiveZone.name].Composition[sendZone.name] = sendZone.deviation
+				receiveZone.deviation -= sendZone.deviation
+				endpointsAvailable.pop()
+				indexDelta++
+			}
+		}
+		endpointsNeeded.pop()
+	}
+	// all endpoints should have been distributed properly. This happens when
+	// the sum of approximated available endpoints > sum of approximated
+	// endpoints in need
+	if len(endpointsAvailable.byZone) != 0 {
+		makeupForApproximation(endpointsAvailable, sliceGroups, true)
+	}
+	return nil
+}
+
+// this helper function helps to deal with situation where the endpoints in need
+// != endpoints available due to the deviation caused by approximation while
+// calculating the expected endpoints of every zone.
+// 1. endpoints needed > endpoints available -> availableOverflow == false,
+// endpoints = list of needed endpoints
+// 2. endpoints available > endpoints needed -> availableOverflow == true,
+// endpoints = list of endpoints available
+func makeupForApproximation(endpoints *endpointsList, sliceGroups map[string]EndpointSliceGroup, availableOverflow bool) {
```

Comment:
If I'm understanding this function correctly, I think we have some room to simplify it. As far as I can tell, it's called in 2 places right now:
1. Endpoints are needed but none are available - noop
2. Endpoints are available but none are needed - simple loop

I think we can get rid of 1 entirely which leaves us with 2. Given that 2 is only ~4 lines of code, I'd be tempted to just embed that into the function above around line 145.

Reply:
done